### PR TITLE
[APP-776] Notify me

### PR DESCRIPTION
**Pull request by phillipsgoncalves**
**What does this PR do?**

This PR aims at implementing the cancel flow for the notify me button on the app coming soon card.

**Database changed?**

   No

**Where should the reviewer start?**

- [ ] BundlesResponseMapper.java

**How should this be manually tested?**

Press the Notify me button, check that you receive the notification. Repeat the flow, but press the cancel button after pressing the notify me button and check that you DO NOT receive the notification.

**What are the relevant tickets?**

  Tickets related to this pull-request: [APP-776](https://aptoide.atlassian.net/browse/APP-776)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new keys, etc.) 

**Code Review Checklist**

- [ ] Documentation on public interfaces
- [ ] Database changed? If yes - Migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] New Kotlin code has unit tests
- [ ] New flows in presenters unit tests
- [ ] Mappers/Validators with any kind of logic unit tests
- [ ] Functional tests pass

### CODE REVIEW

File: app/src/main/java/cm/aptoide/pt/home/HomePresenter.java [hunk 1]
```diff
@@ -704,9 +708,43 @@ private void handleLoadMoreError() {
                   .getCardId());
         })
         .map(HomeEvent::getBundle)
-        .cast(ActionBundle.class)
-        .flatMapCompletable(bundle -> home.setupAppComingSoonNotification(bundle.getActionItem()
-            .getUrl()))
+        .cast(AppComingSoonPromotionalBundle.class)
+        .flatMap(bundle -> home.setupAppComingSoonNotification(bundle.getActionItem()
+            .getUrl())
+            .andThen(Observable.just(bundle)))
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/home/HomePresenter.java:714]: This is minor and up 2 you but I would change this to not force an observable creation with bundle emission when it looks like you want to actually map the emission back to bundle.

```suggestion
        .flatMap(bundle -> home.setupAppComingSoonNotification(bundle.getActionItem()
            .getUrl())
            .toObservable().map(__ -> bundle))
```

We actually want to subscribe to a fire and forget completable without changing the object... like a doOnNext that takes a function... But I don't even know the best way to do that. doOnNext with a completable and .subscribe there is just calling for garbage collection problems... dunno :hammer:

File: app/src/main/java/cm/aptoide/pt/home/HomePresenter.java [hunk 2]
```diff
@@ -704,9 +708,43 @@ private void handleLoadMoreError() {
...
+        .doOnNext(bundle -> {
+          bundle.setRegisteredForNotification(true);
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/home/HomePresenter.java:716]: I don't agree with this manipulation here. There are 3 entities currently manipulating this field:
- The repo/service which creates the object with the default (false)
- This presenter which is changing it for the notify click and cancel notify click.
- The adapter to update the bundle in the view.

We could at least just pass the value down to the adapter and cut the 2nd manipulation which seems a bit unnecessary in my opinion.

File: app/src/main/java/cm/aptoide/pt/home/HomePresenter.java [hunk 3]
```diff
@@ -704,9 +708,43 @@ private void handleLoadMoreError() {
...
+        })
+        .map(HomeEvent::getBundle)
+        .cast(AppComingSoonPromotionalBundle.class)
+        .flatMap(bundle -> home.cancelAppComingSoonNotification(bundle.getActionItem()
+            .getUrl())
+            .andThen(Observable.just(bundle)))
+        .doOnNext(homeBundle -> {
+          homeBundle.setRegisteredForNotification(false);
+          view.updateAppComingSoonStatus(homeBundle);
+        })
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/home/HomePresenter.java:747]: pretty much the exact same comments as the chain above.

File: app/src/main/java/cm/aptoide/pt/home/bundles/BundlesAdapter.java [hunk 4]
```diff
@@ -296,4 +297,21 @@ public boolean isLoaded() {
...
+            .getUrl()
+            .equals(homeBundle.getActionItem()
+                .getUrl())) {
+          int index = bundles.indexOf(bundle);
+          ((AppComingSoonPromotionalBundle) bundle).setRegisteredForNotification(
+              homeBundle.isRegisteredForNotification());
+          notifyItemChanged(index);
+        }
+      }
+    }
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/home/bundles/BundlesAdapter.java:315]: I may be oversimplifying but... can't this be replaced by?:

```suggestion
    int index = bundles.indexOf(homeBundle);
    ((AppComingSoonPromotionalBundle) bundles.get(index)).setRegisteredForNotification(homeBundle.isRegisteredForNotification());
    notifyItemChanged(index);
```

File: app/src/main/java/cm/aptoide/pt/notification/ComingSoonNotificationManager.kt [hunk 5]
```diff
@@ -5,10 +5,12 @@ import android.app.NotificationManager
 import android.content.Context
 import android.os.Build
 import androidx.work.*
+import cm.aptoide.pt.home.AppComingSoonPreferencesManager
 import rx.Completable
 import java.util.concurrent.TimeUnit
 
-class ComingSoonNotificationManager(private val context: Context) {
+class ComingSoonNotificationManager(private val context: Context,
```
Comment (jdandrade) [inline app/src/main/java/cm/aptoide/pt/notification/ComingSoonNotificationManager.kt:12]: This is related to last PR but ideally this class should not have a context and depend on the instance of workmanager and notificationmanager.

**Review by jdandrade — CHANGES_REQUESTED** [review]

File: app/src/main/java/cm/aptoide/pt/home/HomePresenter.java [hunk 6]
```diff
@@ -704,9 +708,43 @@ private void handleLoadMoreError() {
                   .getCardId());
         })
         .map(HomeEvent::getBundle)
-        .cast(ActionBundle.class)
-        .flatMapCompletable(bundle -> home.setupAppComingSoonNotification(bundle.getActionItem()
-            .getUrl()))
+        .cast(AppComingSoonPromotionalBundle.class)
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply app/src/main/java/cm/aptoide/pt/home/HomePresenter.java:711]: I did it, but dont know if it damages the readability, which is the reason why a lot of times i dont like using pairs.

File: app/src/main/java/cm/aptoide/pt/home/HomePresenter.java (hunk 1 above)
Comment (phillipsgoncalves) [reply app/src/main/java/cm/aptoide/pt/home/HomePresenter.java:715]: Tested this and as we spoke on the office, it did not work. The setupNotification method is completable, so it is a completed observable, meaning we can not use doOnNext after, which is what we need.

File: app/src/main/java/cm/aptoide/pt/home/HomePresenter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdandrade) [reply app/src/main/java/cm/aptoide/pt/home/HomePresenter.java:713]: We are still doing the double cast.

**Review by jdandrade — CHANGES_REQUESTED** [review]
Missing migration. Log:
    java.lang.IllegalStateException: Room cannot verify the data integrity. Looks like you've changed schema but forgot to update the version number. You can simply fix this by increasing the version number.

**Review by jdandrade — APPROVED** [review]
lgtm

**Pull request merged**